presto-hive/src/main/java/io/prestosql/plugin/hive/HiveMetadata.java

```diff
@@ -1890,7 +1896,82 @@ private static boolean filterSchema(String schemaName)
     @Override
     public ConnectorTableHandle beginDelete(ConnectorSession session, ConnectorTableHandle tableHandle)
     {
-        throw new PrestoException(NOT_SUPPORTED, "This connector only supports delete where one or more partitions are deleted entirely");
+        HiveTableHandle handle = (HiveTableHandle) tableHandle;
+
+        SchemaTableName tableName = handle.getSchemaTableName();
+        Table table = metastore.getTable(new HiveIdentity(session), tableName.getSchemaName(), tableName.getTableName())
+                .orElseThrow(() -> new TableNotFoundException(tableName));
+        ensureTableSupportsDelete(table);
+
+        long transactionId = metastore.getOrCreateTransactionId().get();
+
+        metastore.acquireTableWriteLock(
+                new HiveIdentity(session),
+                "delete in " + handle.getTableName(),
+                transactionId,
+                handle.getSchemaName(),
+                handle.getTableName(),
+                DataOperationType.DELETE);
+
+        long writeId = metastore.allocateWriteId(handle.getSchemaName(), handle.getTableName(), transactionId);
+
+        HiveTableHandle deleteHandle = handle.forDelete(transactionId, writeId);
+        return deleteHandle;
+    }
+
+    @Override
+    public void finishDelete(ConnectorSession session, ConnectorTableHandle tableHandle, Collection<Slice> fragments)
+    {
+        HiveTableHandle handle = (HiveTableHandle) tableHandle;
+        checkArgument(handle.isDelete(), "handle should be a delete handle, but is " + handle);
+        // The fragments list contains the partitionNames of partitions in which deletion happened, and therefore
+        // the partition's writeId should be updated.
+        requireNonNull(fragments, "fragments is null");
+
+        SchemaTableName tableName = handle.getSchemaTableName();
+        Table table = metastore.getTable(new HiveIdentity(session), tableName.getSchemaName(), tableName.getTableName())
+                .orElseThrow(() -> new TableNotFoundException(tableName));
+
+        ensureTableSupportsDelete(table);
+
+        Set<String> partitionIds = fragments.stream().map(Slice::toStringUtf8).collect(Collectors.toUnmodifiableSet());
+        if (!partitionIds.isEmpty()) {
+            Map<String, Optional<Partition>> partitionsMap = metastore.getPartitionsByNames(
+                    new HiveIdentity(session),
+                    tableName.getSchemaName(),
+                    tableName.getTableName(),
+                    ImmutableList.copyOf(partitionIds));
+            List<Partition> partitions = partitionsMap.values().stream()
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .collect(toImmutableList());
+            if (!partitions.isEmpty()) {
+                metastore.alterPartitions(table.getDatabaseName(), table.getTableName(), partitions, handle.getWriteId());
```

Comment:
> I though about changing the API so that we pass names only, along with write id, down to ThriftHiveMetastore and let it do the update.

That would omit cache right? It seems that we are inconsistent with how we perform Metastore updates. Sometimes we do that kind of atomically without using cache:
* `updatePartitionStatistics`

and sometimes we use cached data:
* `alterPartition`.

It seems if we want to reduce Metastore load for insert operations we should use cached metadata. However, then we might interact with other engines updating metadata at same time. Questions are:
* Do we update sensitive fields? Using cached data for stats seems OK.
* What is the time window between reading from metadata cache and updating partitions in Metastore.
* Can we leave metastore inconsistent when two Presto queries operate at same time? (both use session metadata cache).

I don't have enough context in this PR to answer this questions.

Reply:
I'm not sure I have the answers either.  I'll ping @electrum and see what he says.